Expose the PSA RNG in mbedtls

gilles-peskine-arm opened the pull request:
Expose whatever RNG the PSA subsystem uses to applications using the classic (mbedtls_xxx) API. This is especially useful for applications that use both PSA and TLS and want to leverage the PSA RNG to either save memory or make use of `mbedtls_psa_external_get_random`.

Resolves #3883.

stevew817 commented on include/mbedtls/psa_util.h:
```diff
@@ -419,4 +419,88 @@ static inline int mbedtls_psa_tls_ecpoint_to_psa_ec( unsigned char const *src,
+ *                      `MBEDTLS_ERR_CTR_DRBG_xxx` or
+ *                      `MBEDTLS_ERR_HMAC_DRBG_xxx` on error.
+ */
+int mbedtls_psa_get_random( void *p_rng,
```
Isn't this a straight-up copy of what's in `psa_crypto_random_impl.h`? If this is meant to replace that, why isn't the other file being removed?

gilles-peskine-arm replied:
These are just the declarations that need to be in public headers. `psa_crypto_random_impl.h` has more declarations and definitions.

There's some duplication between the two headers, but I think it's the minimum necessary, or at least the minimum advisable. A couple of things are duplicated, in a way that a compiler will complain if they diverge, and this way all the implementation stuff is in `psa_crypto_random_impl.h` where it's easy to see all in one place, but we still get the compiler to complain if the implementation no longer matches the public interface.

gilles-peskine-arm replied:
Are you convinced by my explanation, or do some things need to be removed from `psa_crypto_random_impl.h`?

gilles-peskine-arm replied:
Once again I'm not sure what change you expect. To me, the comments at the top of the files referring to each other are sufficient. Do you want this to be also mentioned in the documentation of each macro/function/type?

ronald-cron-arm replied:
From your comment above the reason of the duplication seems to be:
"A couple of things are duplicated, in a way that a compiler will complain if they diverge, and this way all the implementation stuff is in psa_crypto_random_impl.h where it's easy to see all in one place, but we still get the compiler to complain if the implementation no longer matches the public interface."
I'm just saying what about to put something equivalent as documentation in the header files.

gilles-peskine-arm replied:
Revisiting this, the compiler doesn't enforce consistency: there's duplication here that's rejected if both headers are included (`static` variable definition, type definition). So I do need to remove some things from `psa_crypto_random_impl.h`.

gilles-peskine-arm replied:
I've removed the redundant definitions and kept redundant declarations where possible. I kept the duplication of the definition of the macro `MBEDTLS_PSA_RANDOM_STATE` because this is both permitted and checked.

stevew817 commented on ChangeLog.d/mbedtls_psa_get_random.txt:
```diff
@@ -0,0 +1,8 @@
+   * Partial implementation of the PSA crypto driver interface: Mbed TLS can
+     now use an external random generator instead of the library's own
+     entropy collection and DRBG code. Enable MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
+     and See the documentatio of mbedtls_psa_external_get_random() for details.
```
typos: `see` and `documentation`

stevew817 commented on ChangeLog.d/mbedtls_psa_get_random.txt:
```diff
@@ -0,0 +1,8 @@
+     entropy collection and DRBG code. Enable MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
+     and See the documentatio of mbedtls_psa_external_get_random() for details.
+   * Applications using PSA crypto can now use its random generator in the
+     mbedtls_xxx API. See the documentation of mbedtls_psa_get_random() for
```
Is it even recommended that applications using PSA Crypto continue to call directly into `mbedtls_xxx`? Having this wording in the public changelog makes it look like it's expected behaviour...

gilles-peskine-arm replied:
This is the opposite: applications using `mbedtls_xxx` features (typically X.509/SSL) and calling into PSA. Is the wording here misleading?

stevew817 replied:
I think the wording might be too generic, in that case. To me (though my headspace is very much in crypto land and not in SSL) it reads like an implied approval to mix-and-match PSA and mbedtls_xxx for cryptographic functionality.

gilles-peskine-arm replied:
I reworded the second changelog entry. Is it clear now?

stevew817 reviewed (changes requested):
Overall, this PR looks good to me. It adds testing coverage to the work from #3895 and defines usage paradigms for the exposed functionality.

Just a couple of stylistic remarks.

ronald-cron-arm reviewed (changes requested):
This looks almost good to me. I have a few comments though.

gilles-peskine-arm commented on programs/ssl/ssl_test_lib.c:
```diff
@@ -84,6 +92,18 @@ int rng_seed( rng_context_t *rng, int reproducible, const char *pers )
         return( -1 );
     }
 #endif
+#if defined(MBEDTLS_TEST_USE_PSA_CRYPTO_RNG)
```
This implication is somewhat coincidental. Furthermore, `MBEDTLS_USE_PSA_CRYPTO` itself is not intrinsically incompatible with reproducible mode: it could be made conditionally compatible based on the PSA RNG being compatible, in which case the first block could be removed but not the second one.

ronald-cron-arm replied:
Yes it makes sense.

gilles-peskine-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -3035,6 +3022,20 @@ int main( int argc, char *argv[] )
 #endif /* MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED &&
           MBEDTLS_USE_PSA_CRYPTO */
 
+#endif
```
Moving the `#endif` was a mistake. A comment would have made it more apparent, I'll add one.

gilles-peskine-arm commented on programs/ssl/ssl_test_lib.h:
```diff
@@ -43,23 +43,33 @@
+#undef HAVE_RNG
+#if defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG) &&         \
+    ( defined(MBEDTLS_USE_PSA_CRYPTO) ||                \
+      defined(MBEDTLS_TEST_USE_PSA_CRYPTO_RNG) )
```
As documented in a comment below, the implication is somewhat coincidental. Furthermore the implication is not true yet at this point (sure, the file could be restructured accordingly, but I think the current ordering makes more sense).

gilles-peskine-arm commented:
I modified the history in the following way (previous history in https://github.com/gilles-peskine-arm/mbedtls/tree/psa-external-random-in-mbedtls-4):

* New commit "Add comments to a few #endif closing a long chunk".
* Fixed https://github.com/ARMmbed/mbedtls/pull/4110#discussion_r576747676 by amending "Fix psa_destroy_key called after mbedtls_psa_crypto_free".
* Squashed the copypasta fix for https://github.com/ARMmbed/mbedtls/pull/4110#discussion_r576701456 into the commit that introduced it.

stevew817 reviewed (approved):
Nothing more from my side.

d3zd3z reviewed (approved):
Looks good to me.

daverodgman reviewed (approved):
Seems that all of Ronald's comments have been addressed. Merging as Ronald is out this week.
